blogposts/2020/optimizing-a-code-intelligence-indexer.md

```diff
@@ -0,0 +1,196 @@
+
+```go
+func ensureValue(key string, f func() string) string {
+    // fast path
+    mutex.RLock()
+    value, ok := m[key]
+    mutex.RUnlock()
+    if !ok {
+        return value
+    }
+
+    // slow path
+    mutex.Lock()
+    defer mutex.Unlock()
+
+    if value, ok := m[key]; ok{
+        return value
+    }
+
+    // note: f() can be moved outside of critical
+    // section if it's expensive/side-effect free
+    value = f()
+    m[key] = value
+    return value
+}
+```
+
+This pattern first acquires a cheap read lock on `mutex` and checks the likely condition that the data has already been cached in the shared map `m`. If the value exists, it can be returned. Otherwise, a value needs to be generated and written to the shared map for later readers. In order to handle the race condition between releasing the read lock and acquiring the write lock, we check for the value in the map again after exclusive access has been acquired.
+
+Each shared datastructure has its own mutex so that contention from one resource does not bleed into a less-contended resource. However, it's easy to take this pattern too far. We have a shared two-layer hierarchical map whose access is protected by a [striped mutex](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html). A striped mutex is basically an array of shared mutexes, where a resource `r` is locked if the mutex at index `hash(r)` is locked (modulo the configured number of mutexes). This is a good middle-ground between having one mutex for a highly contended resource, and having `n` distinct mutexes for a map with `n` keys. The former will decrease the throughput of of the indexer as every routine is waiting on the same lock, and the latter will blow up memory usage, as `n` can be very large even for relatively small code bases.
+
+##### You have to be sure that there are no hidden data dependencies
+
+When we index references, we assume that our map of definitions is already populated. Our solution is to parallelize the workload in distinct steps (first index all definitions, then index all references), where each step can rely on data calculated by hte previous step which was parallelized individually.
+
+Alternative solutions here require pushing back tasks whose dependencies have not yet been met, pre-determining a dependency graph, or recursively submitting duplicate work to the queue. Each of these solutions have a down side and decrease code clarity.
+
+##### You have to choose the right level of work to parallelize
+
+You can't throw work into a parallel executor and guarantee that your program runs faster. In fact, the first few attempts at parallelizing lsif-go wasn't fruitful because we tried to index each _file_ in parallel rather than each _package_. The cost of indexing a single file was low enough that the overhead of setting up the goroutines and queueing the work negated any benefit of executing the work in multiple cores.
```

Comment:
```suggestion
You can't throw work into a parallel executor and guarantee that your program runs faster. In fact, the first few attempts at parallelizing lsif-go weren't fruitful because we tried to index each _file_ in parallel rather than each _package_. The cost of indexing a single file was low enough that the overhead of setting up the goroutines and queueing the work negated any benefit of executing the work in multiple cores.
```